### ISSUE TITLE
Publish to PyPI from GHA with PEP 740 digital attestations

### ISSUE THREAD

**Issue by webknjaz** [adrienverge/yamllint#721]
This would allow showing more metadata as verified on PyPI and include verifiable dist signatures uploaded there too.

Here's the guide to follow: https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/. By implementing it, the digital attestations will be included automatically. Ping me for review if need be.

**Comment by adrienverge** [comment]
@webknjaz thanks, it's a good idea! I implemented this at https://github.com/adrienverge/yamllint/pull/724.

I'll merge now and create a new tag to make sure it works on the real PyPI, but feel free to review the pull request: we can make edits afterwards if needed.

**Comment by webknjaz** [comment]
@adrienverge looks like you'll need to set up a trusted publisher on TestPyPI or drop that job. Although, if you want to keep it for regular PR merges, you need some extra version computation for untagged commits since (TestPyPI) won't allow version reuse. I usually use `setuptools-scm` for this.

**Comment by adrienverge** [comment]
I did set up a trusted publisher on TestPyPI (same procedure as on PyPI) prior to pushing anything on the GitHub repo, so it looks like the HTTP 400 sometimes returned by test.pypi.org are rather due to duplication of pushes, as your comment suggests.

I want to fix this but I see a benefit in keeping these test publications to TestPyPI, as regular sanity checks.

To avoid pushing releases with the same version, we could either:
- run the whole worflow only if on master branch:
  ```yaml
  on:
    push:
      branches:
        - master
  ```
- run the workflow on all pushes, but the TesPyPI job only on master branch:
  ```yaml
    publish-to-testpypi:
      if: github.ref == 'refs/heads/master'
  ```

What do you think?
Is there a preferred option? An alternative option? Are there best practices?

**Comment by adrienverge** [comment]
In the meantime I added `skip-existing: true` on TestPyPI to prevent these HTTP 400s, see https://github.com/adrienverge/yamllint/pull/727.

**Comment by adrienverge** [comment]
Thanks for all these details Sviatoslav, it's interesting and valuable.

Here, the workflow can remain simple so I'd like to avoid variables (I understand they're useful in more complex cases like some of yours) and we can assume there won't be tags on anything else than on release commits.
And like most repo maintainers, I'd like tags to trigger publications, despite the drawback you pointed out. Less work (as long as everything goes well), easier to understand.

So in the end, I propose to simply change:

```diff
--- a/.github/workflows/publish.yaml
+++ b/.github/workflows/publish.yaml
@@ -29,6 +29,7 @@ jobs:
 
   publish-to-testpypi:
     name: Publish distribution package to TestPyPI
+    if: github.ref == 'refs/heads/master'
     needs: build
     runs-on: ubuntu-latest
     environment:
@@ -46,7 +47,6 @@ jobs:
         uses: pypa/gh-action-pypi-publish@release/v1
         with:
           repository-url: https://test.pypi.org/legacy/
-          skip-existing: true
```

About `twine check`, I [read that](https://twine.readthedocs.io/en/stable/#twine-check) it just "*checks whether your distribution’s long description will render correctly on PyPI*". So is it really useful? Does it check anything else?

```diff
--- a/.github/workflows/publish.yaml
+++ b/.github/workflows/publish.yaml
@@ -21,6 +21,8 @@ jobs:
         run: python -m pip install --user build
       - name: Build a binary wheel and a source tarball
         run: python -m build
+      - name: Twine check the distribution packages
+        run: twine check --strict dist/yamllint-*.tar.gz dist/yamllint-*.whl
       - name: Store the distribution packages
         uses: actions/upload-artifact@v4
```

I'll wait for your response before changing anything ;)

**Comment by webknjaz** [comment]
> Thanks for all these details Sviatoslav, it's interesting and valuable.
> 
> Here, the workflow can remain simple so I'd like to avoid variables (I understand they're useful in more complex cases like some of yours) and we can assume there won't be tags on anything else than on release commits. And like most repo maintainers, I'd like tags to trigger publications, despite the drawback you pointed out. Less work (as long as everything goes well), easier to understand.
> 

Sure, I just wanted to document the options explicitly.

> So in the end, I propose to simply change:
> 
> ```diff
> --- a/.github/workflows/publish.yaml
> +++ b/.github/workflows/publish.yaml
> @@ -29,6 +29,7 @@ jobs:
>  
>    publish-to-testpypi:
>      name: Publish distribution package to TestPyPI
> +    if: github.ref == 'refs/heads/master'
>      needs: build
>      runs-on: ubuntu-latest
>      environment:
> @@ -46,7 +47,6 @@ jobs:
>          uses: pypa/gh-action-pypi-publish@release/v1
>          with:
>            repository-url: https://test.pypi.org/legacy/
> -          skip-existing: true
> ```

This will work when you merge the first PR after a release and will fail when you merge another one as long as the version is hardcoded, which it is. It can only work if you ensure that each one of those builds ends up having a unique version number. I use `setuptools-scm` for this. You could theoretically  `sed` that Python file in such cases if you want to re-hardcode stuff w/o actually committing. But you'd need to be careful to only do this on non-release requests.

> About `twine check`, I [read that](https://twine.readthedocs.io/en/stable/#twine-check) it just "_checks whether your distribution’s long description will render correctly on PyPI_". So is it really useful? Does it check anything else?

It might start checking for various metadata incompatibilities in the future, though. But more importantly is that `pypi-publish` runs this check too. And if you don't run in continuously, you may hit an error too late, close to the end of the release workflow. This is usually unpleasant.

> ```diff
> --- a/.github/workflows/publish.yaml
> +++ b/.github/workflows/publish.yaml
> @@ -21,6 +21,8 @@ jobs:
>          run: python -m pip install --user build
>        - name: Build a binary wheel and a source tarball
>          run: python -m build
> +      - name: Twine check the distribution packages
> +        run: twine check --strict dist/yamllint-*.tar.gz dist/yamllint-*.whl
>        - name: Store the distribution packages
>          uses: actions/upload-artifact@v4
> ```
> 
> I'll wait for your response before changing anything ;)

Since we know that there's just 2 files expected there at all times, doing `python -Im twine check --strict dist/*` would be enough.

**Comment by adrienverge** [comment]
> This will work when you merge the first PR after a release and will fail when you merge another one as long as the version is hardcoded, which it is.

Of course, thanks. So I'll use `sed`.
Please see https://github.com/adrienverge/yamllint/pull/728.
